### ISSUE TITLE
Close java.lang.AutoCloseable resources returned by DataFetcher

### ISSUE THREAD

**Issue by asauvez** [graphql-java/graphql-java#1863]
### Explanation
One of my `DataFetcher` returns a wrapper around a sort of `java.sql.ResultSet`. My wrapper implements `java.lang.AutoCloseable` in order to free resources once the query is done.

### Current behavior
Currently, resources are not released when the request ends. 

### Proposed/Desired behavior
If the object returned by the `DataFetcher` implements `java.lang.AutoCloseable`, the method `close()` is called on it  at some point before the query end.

**Comment by bbakerman** [comment]
So its a completely crazy idea.... le is think on it.

But you can kinda do this today via CompletableFuture.  The DataFecthers in graphql-java handle a number of different types of objects and CompletableFuture is one of them

So given any value you could do this today

```
DataFecther df = environment -> {
      Object value = /// get value some how
      CompletableFuture cf = CompletableFuture.completedValue(value);

       return cf.thenApply (v -> {
             if (v instance of Closeable) {
                    ((Closerable) v).close()
            }
      }
      });
}
```

This will close the value when the completable future finishes.  Now this could be generalised in a help function that also checks if `value` is a cf and not wrap it.  etc..

My point is you CAN do this today via a small amount of code.

And in fact via `graphql.execution.instrumentation.Instrumentation#instrumentDataFetcher` you can turn ANY data fetcher into one that does this CF / Closeable checking without changing any lower level data fetcher code

**Comment by bbakerman** [comment]
eg

```
    static CompletableFuture<?> closeableSupport(Object value) {
        CompletableFuture<?> cf;
        if (value instanceof CompletionStage) {
            cf = ((CompletionStage<?>) value).toCompletableFuture();
        } else {
            cf = CompletableFuture.completedFuture(value);
        }

        return cf.thenAccept(val -> {
            if (val instanceof AutoCloseable) {
                try {
                    ((AutoCloseable) val).close();
                } catch (Exception e) {
                    // decide your strategy here
                }
            }
        });
    }

```

Then in the DF 

```
                DataFetcher<?> df = environment -> closeableSupport(someValue);
```

**Comment by asauvez** [comment]
With this code, `close()` is called before the resultSet is used:

In your first code, since the `CompletableFuture` is already completed, the `thenApply` is applied even before the end of the `DataFetcher`.

**Comment by bbakerman** [comment]
Great point - you cant use the code above.  It would close it before it was processed

Scratch that idea.

**Comment by dondonz** [comment]
Hello as this issue is quite old and hasn't received any extra comments lately, I'll close this issue out.

If you'd like to make a new comment please do open a new issue, thanks!
